cpp/src/arrow/util/async_generator.h

```diff
@@ -775,6 +791,7 @@ class ReadaheadGenerator {
     if (state_->finished.load()) {
       state_->readahead_queue.push(AsyncGeneratorEnd<T>());
     } else {
+      state_->num_running.fetch_add(1);
```

Comment:
This seems off; that means on the initial call, we'll set num_running to max_readahead + 1?

Reply:
Though I guess we're comparing with 1 instead of 0 up above…'num_running' doesn't sound like quite the right name to me but it's also unclear how exactly to articulate what exactly this counter represents.

Reply:
The comparison with `1` is a little misleading.  `fetch_sub` returns the value before the modification so we are testing whether we were the one to set it to 0.

`max_readahead + 1` seems right to me.  `max_readahead` is the number of "extra spots" so if a user asks for 1 item and they have 4 readahead we need to start 5 requests, one to go straight to the caller and the other 4 to fill up the readahead spots.

Why does `num_running` seem like the wrong name?  If this is, for example, file readahead, and num_running is 5 then we are reading from 5 different files.

Reply:
Ah, ok, I missed the fact that `max_readahead` is extra spots so I got confused.